### PR TITLE
ENG-7829 -  iOS Throttling

### PR DESCRIPTION
**Pull request by kevinsites**
This PR supports remote throttling via the mobile config. The remote config will be loaded on a starting command (`start/startSession/startAppFlow`), then depending if throttling is required the session/flow will be determined to be "sampled" or not. If the session/flow should NOT be sampled then listeners will still be attached but events will drop once they reach the data store. 

NOTE: In order to fully support this remote config without a race condition a re-work was required for all starting commands `start/startSession/startAppFlow` to take a completion callback. This is because async/await is not supported in swift until iOS 15 and we support back to iOS 12. 
Loading of config was also taken OUT of the `configure` command to lessen the number of command signatures changed (I.e. instead of adding a callback to configure AND starting commands, since we have to wait at start regardless I just put it there). 
The listener/service creation piece that was happening in the `configure` command was put into its own function to be called via the starting command. 

Additionally this PR includes:
- Rework of the NIDConfigureService to be used as an attribute of NeuroID instead of static class everywhere, new caching methods to check and update the config, new method to update the config options based on siteID passed in
- Update to our network service/mock service to include a generic GET request
- Reorganizing some class attributes to group like things together
- A rework of the inner pieces of start/startSession to call a single command `setupSession` that sets up the session (the code was already 99% the same with a difference in the initTimer functions called) including determining sampling rate, attaching listeners/creating the classes, and the other necessary things.

Additional Note:
I completely commented out the `NIDPerformanceTests` because they seemed to be unstable and always caused other tests to fail because they were running concurrently and changing the DataStore.events count/content. Once these tests are more stable and/or once we move to dependency injection instead of static classes this can be fixed.

### CODE REVIEW

**Comment by kevinsites** [comment]
For reference the new starting commands will look like this:

`
NeuroID.start{ started in
 // callback completion here
}

// ADV
NeuroID.start(true){ started in
 // callback completion here
}

NeuroID.startSession( "myUserID"){ started in
 // callback completion here
}

NeuroID.startSession(nil){ started in
 // callback completion here
}

// ADV
NeuroID.startSession( "myUserID", true){ started in
 // callback completion here
}

NeuroID.startSession(nil, true){ started in
 // callback completion here
}


NeuroID.startAppFlow("mySite"){ started in
 // callback completion here
}

NeuroID.startAppFlow("mySite", "myUserID"){ started in
 // callback completion here
}


`

File: NeuroID/NeuroIDClass/NeuroID.swift
Comment (jeff-niemann) [inline NeuroID/NeuroIDClass/NeuroID.swift:60]: Thoughts on `isSessionFlowSampled`? Since our primary use case is on the sessionFlow

File: NeuroID/Utilities/NIDConfigService.swift
Comment (jeff-niemann) [inline NeuroID/Utilities/NIDConfigService.swift:83]: minor nit: `siteID == nil || siteID ?? "" == self.configCache.siteID ?? "noID"` could be written as a function `isCollectionSite(siteId)` to make this more readable

File: NeuroID/NeuroIDClass/Extensions/NIDAdvancedDevice.swift
Comment (jeff-niemann) [inline NeuroID/NeuroIDClass/Extensions/NIDAdvancedDevice.swift:27]: Are there any TODOs that need to be called out with this comment?

Since we are changing the interface, and doing a rev of the SDK, is backwards compatibility needed here?

File: NeuroID/NeuroIDClass/Extensions/NIDSession.swift
Comment (jeff-niemann) [inline NeuroID/NeuroIDClass/Extensions/NIDSession.swift:214]: If sampling is > 100, then we should receive all traffic.. right?

File: NeuroID/NeuroIDClass/Extensions/NIDSession.swift
Comment (jeff-niemann) [inline NeuroID/NeuroIDClass/Extensions/NIDSession.swift:212]: Minor nit: personally I prefer `updateIsSampledStatus` or something that is more descriptive of the behavior of _how_  `isSessionSampled` is used throughout the code.

The pattern I have noticed is read `isSessionSampled` and we update `isSessionSampled` at logical points

File: NeuroID/Utilities/NIDConfigService.swift
Comment (jeff-niemann) [inline NeuroID/Utilities/NIDConfigService.swift:92]: Personally I would prefer a more immutable approach to the configCache. Ideally the configCache just holds sample rates by siteId/linkedSiteId, and does not have to do any logic to resolve "what the flow the user is currently in" 

Then in the `determineIsSessionSampled` function, we resolve the siteId to the available sites in the cache to find the sample rate.

This approach will **decouple** this logic from the simple cache, and push the business logic to where it belongs (`determineIsSessionSampled`)

**Comment by kishisaka-nid** [comment]
Due to the completion callback refactor, it looks like we will have to change out the ReactNative wrapper to support these changes. Is this the case?

File: NeuroID/NeuroIDClass/Extensions/NIDAdvancedDevice.swift
Comment (kevinsites) [reply NeuroID/NeuroIDClass/Extensions/NIDAdvancedDevice.swift:27]: This is more of an internal note for us to decide before merging. Ed and I were going back and forth on if its needed because its a minor rev

File: NeuroID/NeuroIDClass/Extensions/NIDSession.swift
Comment (kevinsites) [reply NeuroID/NeuroIDClass/Extensions/NIDSession.swift:159]: Correct, there was a slip through where if it equaled 100 then it wouldn't collect. But if I put <= then if it was 0 it would still collect if the rolled value was 0. This is just to double check

File: NeuroID/Utilities/NIDConfigService.swift
Comment (kevinsites) [reply NeuroID/Utilities/NIDConfigService.swift:92]: I thought about that approach however because linkedSites have the opportunity to have more options than just `sample_rate` I did not think `determineIsSessionSampled` was the right spot.

IMO the `determineIsSessionSampled`  function is just to use the config value and determine true/false. It's not to try and get an entire linkedSiteOptions and use them. This function however is built to update the config values based on the site/flow we are using.

I do agree we could make the configCache more immutable and maybe have a temporary holding config that represents the options for the linkedSite, however that seemed more appropriate when we expand the linkedSiteOptions than right now.

**Comment by kevinsites** [comment]
> Due to the completion callback refactor, it looks like we will have to change our the ReactNative wrapper to support these changes. It also looks like we will need to update Android to have this same behavior. Is this the case?

@kishisaka-nid  - That is correct, I am assuming we will have to have a similar update in Android and by extension RN as well. This does not have to be done in this ticket as this is iOS only for the time being

**Review by cometman — CHANGES_REQUESTED** [review]

File: NeuroID/Utilities/NIDConfigService.swift
Comment (jeff-niemann) [reply NeuroID/Utilities/NIDConfigService.swift:92]: `NidConfigService` operates as a caching layer for configs/linkedConfigs applicable to that site.

Multiple service can consume those configs for their own purposes, and there could be many different options within the configs. The service consuming the config should have the business logic of "what it needs to do"

Determining the sample rate is business logic that exist in the service that is designated to determine the sample rate, even if its required to pull all the sample rates of the linkedSiteIds.

I feel this is a very clear distinction, (think single responsibility principle for your services), making the code easier to read and less prone to bugs longer term.

**Comment by kevinsites** [comment]
> pulled, running in swift UI, Looks like the sited is nil in the log. Called:
> 
> `NeuroID.startAppFlow(siteID: "flow_zzzzz333")`
> 
> ```
> (NeuroID Info)  ["NeuroID Response 200"]
> (NeuroID Info)  ["NeuroID Payload: NeuroHTTPRequest(clientId: \"BFE73E05-253E-48CB-8250-7D5819F1AC68\", environment: \"LIVE\", sdkVersion: \"5.ios-3.2.0\", pageTag: \"LoanInfo\", responseId: \"3E81D002\", siteId: \"\", userId: Optional(\"Fsfsdf-swiftui-sandbox\"), registeredUserId: nil, jsonEvents: [NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent], tabId: \"463DA00933D6\", pageId: \"244A542F5E9D\", url: \"ios://LoanInfo\", jsVersion: \"5.0.0\", linkedSiteId: nil)"]
> ```

The function will not set the `siteID`, it will set the `linkedSiteId`, however your siteID is invalid (check your logs or the callback response object). You used `flow_XXX` and it validates against `form_xxx`

**Comment by kishisaka-nid** [comment]
pulled and run with a proper form name this time. Looks good: 

`NeuroID.startAppFlow(siteID: "form_zzzzz333")`
```
(NeuroID Info)  ["NeuroID Response 200"]
(NeuroID Info)  ["NeuroID Payload: NeuroHTTPRequest(clientId: \"BFE73E05-253E-48CB-8250-7D5819F1AC68\", environment: \"LIVE\", sdkVersion: \"5.ios-3.2.0\", pageTag: \"PersonalInfo\", responseId: \"56D76002\", siteId: \"\", userId: Optional(\"Fsdadfsdaf-swiftui-sandbox\"), registeredUserId: nil, jsonEvents: [NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent, NeuroID.NIDEvent], tabId: \"2D5FCD7D27D6\", pageId: \"0A1C14CE135B\", url: \"ios://PersonalInfo\", jsVersion: \"5.0.0\", linkedSiteId: Optional(\"form_zzzzz333\"))"]
```

File: NeuroID/Utilities/NIDConfigService.swift
Comment (jeff-niemann) [reply NeuroID/Utilities/NIDConfigService.swift:92]: For example: if this service is async refreshing the cache in the background (this is an assumption but generally how a cache works) the service can refresh the cache _without_ calling this business logic.

File: NeuroID/NeuroIDClass/Extensions/NIDSession.swift
Comment (jeff-niemann) [inline NeuroID/NeuroIDClass/Extensions/NIDSession.swift:160]: So if its above 100? we are going to sample it?

File: NeuroID/Utilities/NIDConfigService.swift
Comment (jeff-niemann) [inline NeuroID/Utilities/NIDConfigService.swift:68]: We should consider sending a `SET_VARIABLE` event when the config is refreshed with the config version, so that we can later know what config was applied during collection

**Comment by kishisaka-nid** [comment]
LGTM, 🚢

**Review by kishisaka-nid — DISMISSED** [review]

File: NeuroID/NeuroIDClass/Extensions/NIDSession.swift
Comment (kishisaka-nid) [inline NeuroID/NeuroIDClass/Extensions/NIDSession.swift:95]: Should we consider reseting the linkedSiteId to "" here (or in the form name check) in case of a bad form name which may cause the previous name to be used in the next flow?

Comment (kevinsites) [reply NeuroID/NeuroIDClass/Extensions/NIDSession.swift:95]: That's a good callout, lets discuss post standup as a team on the behavior

File: NeuroID/NeuroIDClass/Extensions/NIDSession.swift
Comment (kevinsites) [reply NeuroID/NeuroIDClass/Extensions/NIDSession.swift:160]: Yes we are going to sample (sample = collect) if it's 100 or higher. Again this was to catch a weird edge case where if the value equaled the min/max it wouldn't behave as expected

File: NeuroID/Utilities/NIDConfigService.swift
Comment (kevinsites) [reply NeuroID/Utilities/NIDConfigService.swift:68]: That's a good idea, we don't currently have a concept of a config "version" but its something we can add a log saying the config was refreshed and then add a version to the config at a later point

File: NeuroID/Utilities/NIDConfigService.swift
Comment (kevinsites) [reply NeuroID/Utilities/NIDConfigService.swift:92]: I understand what you are suggesting, I think we should discuss post standup/on zoom because I think this approach will result in more duplicate code

File: NeuroID/Utilities/NIDConfigService.swift
Comment (jeff-niemann) [reply NeuroID/Utilities/NIDConfigService.swift:68]: We should at least encode the current config into the event. Ive spent way too much time debugging web sessions where the config didnt match what we _THOUGHT_ the config was only to find out it wasnt, and that was causing the bug.

**Review by kishisaka-nid — APPROVED** [review]

**Review by quacke-nid — COMMENTED** [review]
comments

**Review by quacke-nid — APPROVED** [review]

File: NeuroID/NeuroIDClass/Extensions/NIDSession.swift
Comment (kevinsites) [reply NeuroID/NeuroIDClass/Extensions/NIDSession.swift:101]: if they try to set an invalid linkedSiteID then we clear the existing linkedSite and revert back to the parent and then end the function (we don't do the follow up work of CREATE_SESSION, or ADV. We also return their completion handler with a `false` in the SessionStartResult indicating we did not actually start the new flow. If this was to happen for the very first call (no session started) then we would not start anything either

**Review by jeff-niemann — APPROVED** [review]
Things looking good e2e with this branch

File: NeuroID/NeuroIDClass/Extensions/NIDAdvancedDevice.swift
Comment (kevinsites) [reply NeuroID/NeuroIDClass/Extensions/NIDAdvancedDevice.swift]: It was decided to remove the returned value and only use the callback. This is a breaking change for customers however it is a minor bump so breakage is allowed. Additionally this will prevent compiling so a customer cannot ship without changing if they update to the latest version.

**Review by quacke-nid — APPROVED** [review]

**Review by kishisaka-nid — APPROVED** [review]

**Review by jeff-niemann — APPROVED** [review]

**Review by cometman — APPROVED** [review]

**Pull request merged**
